### ISSUE TITLE
no file search?

### ISSUE THREAD

**Issue by rxng** [KELiON/cerebro#115]
is this a planned feature or is there a plugin?

**Comment by KELiON** [comment]
@rxng hey! now there is only file navigation plugin (start input with `/` or `~` to navigate in your file system). 

Regarding files search, I'm trying to investigate and find better ways to do it. Can you now describe your certain case? What kind of files do you usually need to find?

**Comment by rxng** [comment]
I use spotlight and do not think I could live without it! MacOS has mdfind and locate commands so I think it would not be too difficult to integrate this into Cerebro, as these are the fastest search tools in Mac? Awesome job btw so far, tried both this and zazu and definitely prefer the ease of use of Cerebro.

**Comment by Lampwick** [comment]
I'd love file search for Linux too. It's just more convenient than having to type out and remember the full file path.

**Comment by Kageetai** [comment]
I thin a full file search of the full system is a bit much (if not using something like *Everywhere* on Windows (WIP)), but maybe we can also just add custom directories that the user wants to be searched, like the home dir etc.? I'd personally also prefer that to just search the whole system and get a lot of clutter in the search...

**Comment by rxng** [comment]
Yes... alfred has this feature that allows you to specify folders to search in

**Comment by dynobo** [comment]
I also would love to see file search in this great App. My use case is: 
* open the files faster than going through file explorer by
* searching in specified folders recursively for files of specified file types and by
* searching for folder name and contents (for text files at least)
I want to add: This is my mostly missed feature in cerebro and the reason why I probably stick to Albert...

**Comment by masterperas** [comment]
I made a small plugin that uses the Everything cli 
https://github.com/masterperas/es-everything-plugin

**Comment by stevenpack** [comment]
This is an important feature to get MacOS/OSX people to Linux. With spotlight, you can do "Command-Space" and just start typing and files will be included in the search results. It's a great feature for productivity. Yes, there is work to do around performance, indexing, exclusions etc. but it works fantastically well on Mac. It's one of the main things I miss when I"m on my Linux boxes.

**Comment by bmix** [comment]
I have a ton of developer docs stored on my machine. APIs, SDKs and stuff. Usually I start them by entering a few chars of the filename in [Launchy](https://www.launchy.net/) and up it comes. It does this by letting the user define filesystem paths along with filetypes (by suffix) to be indexed and runs them like a double-click action on the file.

An interesting idea, to take this further, as mentioned by others, would be to use the `locatedb` datbase on unixoid systems and to use [Everything](https://www.voidtools.com/) (it can be run as http-server or also using its own [FTP-derived protocol](https://www.voidtools.com/forum/viewtopic.php?t=1790)) on Windows.

To sum it up, I would make this three-fold:
1. giving users the possibility to set up fixed paths, that shall be searched for certain filetypes (builtin)
2. write a plugin, that uses locatedb
3. write a plugin that uses Everything

**Comment by leafonsword** [comment]
Hope support file text serarch~

**Comment by coaxial1** [comment]
I'm also looking forward to see this essential feature integrated! Where can I upvote this? :)

**Comment by brpaz** [comment]
Hello. Here is my plugin for searching using tracker on linux. (https://github.com/brpaz/cerebro-goto-folder-linux)

Its only folder search for now, but can be easily modified to allow file search also.

**Comment by ProbonoBonobo** [comment]
Recently switched from macOS to Ubuntu. I think @stevenpack is exactly right: the absence of a fast, comprehensive filesearch utility _a _ _la_ Spotlight/Alfred is a real blow to my productivity. I also concur with @bmix that `locatedb` looks extremely promising as a drop-in replacement for `mdfind`, which is just macOS's command-line version of Spotlight. Spotlight might even call `mdfind` in its filesearch implementation, I'm not sure -- although if that is the case, it must implement some elementary scoring function to rank the results in descending order of relevance/likelihood, because Spotlight typically yields only a small subset of the results returned by mdfind for the same query (with a strong bias towards files accessed recently).

@brpaz's plugin is a step in the right direction, but it doesn't return files -- only directories. I might see if I can modify it to search files, though. :)

An interesting, but way more challenging, approach might be to sidestep `locatedb` altogether and instead train a doc2vec-style algorithm to create a high-dimensional embedding of the files in the user's filesystem (at least the non-binary ones). Alternatively, Elasticsearch implements a [similarity search indexing module](https://www.elastic.co/guide/en/elasticsearch/reference/current/index-modules-similarity.html) already which works a little differently, but exposes some scoring functions like TF/IDF that are closely related to word2vec/doc2vec embeddings. I'm happy to chat more about this if anyone's interested.

**Comment by Spandan-Madan** [comment]
I am a researcher with some experience with Document Classification and NLP. If someone is interested in building a tool which indexes files based on what @ProbonoBonobo  mentioned above. i'm more than happy to have a skype/phone call to point them in the right direction on how to use this stuff!

**Comment by pascalwhoop** [comment]
I'd like to discuss this again. @Spandan-Madan @KELiON what about using `tracker` which is the backend behind the gnome dash search. It seems to work fine, although it may require careful tweaking to avoid [these issues](https://www.google.com/search?ei=jsVpXJK1DMHUwAKi84HIDg&q=tracker-miner-fs+cpu+usage&oq=tracker-miner-fs+cpu+usage).

A search on my machine for these terms results in this on the tools CLI

```
 tracker search "linux search full text indexed"                                                                                                           172ms  So 17 Feb 2019 21:36:42 CET
Results:
  file:///home/pascalwhoop/Documents/Syncthing/Default%20Folder/_ALL/EBooks/Non-Fiction/Tech/head_first_iphone_development.pdf
  ...pr when the button is decisionText.text = @”Go for it!”; } - (void)dealloc...

  file:///home/pascalwhoop/Documents/Syncthing/Default%20Folder/_ALL/EBooks/Non-Fiction/Tech/HumbleBundle/bashpocketreference_2ndedition.pdf
  ...incr is an integer. The shell expands this construct to the full...

  file:///home/pascalwhoop/Documents/Syncthing/Default%20Folder/_ALL/EBooks/Non-Fiction/Tech/HumbleBundle/bashcookbook.pdf
  ...Search Getting a Simple True/False from a Search Searching for Text...

  file:///home/pascalwhoop/Documents/Syncthing/Default%20Folder/transfer/Linux%20Journal/2016/Linux%20Journal%20-%2003.2016%20-%20Use%20an%20initrd.pdf
  ...PHOTO OF THE MONTH Remember, send your Linux...

  file:///home/pascalwhoop/Documents/Syncthing/Default%20Folder/transfer/Linux%20Journal/2013/Linux%20Journal%2011.2013%20-%20System%20Administration.pdf
  ™ SPONSORED BY Since 1994: The Original Magazine of the Linux Community...

  file:///home/pascalwhoop/Documents/Syncthing/Default%20Folder/transfer/Linux%20Journal/2013/Linux%20Journal%2003.2013%20-%20Web%20Development.pdf
  ...search button next to it. If I enter text in that text...

  file:///home/pascalwhoop/Documents/Syncthing/Default%20Folder/transfer/Linux%20Journal/2012/Linux%20Journal%2009.2012%20-%20Embedded.pdf
  ...full of things for every flavor of Linux enthusiast...

  file:///home/pascalwhoop/Documents/Syncthing/Default%20Folder/transfer/Linux%20Journal/2012/Linux%20Journal%2008.2012%20-%20Linux%20Means%20Business.pdf
  ...nice to have full-text search on all the issues...

  file:///home/pascalwhoop/Documents/Syncthing/Default%20Folder/transfer/Linux%20Journal/2012/Linux%20Journal%2002.2012%20-%20Web%20Development.pdf
  ...files have been indexed by search engines, improved...

  file:///home/pascalwhoop/Documents/Syncthing/Default%20Folder/transfer/Linux%20Journal/2011/Linux%20Journal%2010.2011%20-%20Networking.pdf
  ...Intel’s 320 SSD Line, p. 54 LINUX JOURNAL...


  ...

NOTE: Limit was reached, there are more items in the database not listed here
```

**Comment by lucky125111** [comment]
I was using Everything on Windows, but as I switch to Linux I can't seem to find a good alternative.
I recently created plugin for searching files under Linux [cerebro-search-linux](https://github.com/lucky125111/cerebro-search-linux). 
It used `locate` underneath for instant search results. The trade off is sometimes a new file might not be present in the search results if the db hasn't been updated.
Of course any feedback is highly appreciated.

**Comment by jreus** [comment]
> I was using Everything on Windows, but as I switch to Linux I can't seem to find a good alternative.
> I recently created plugin for searching files under Linux [cerebro-file-search](https://github.com/lucky125111/cerebro-file-search).
> It used `locate` underneath for instant search results. The trade off is sometimes a new file might not be present in the search results if the db hasn't been updated.
> Of course any feedback is highly appreciated.

hey @lucky125111 ~~ this looks great! I'd like to try it out, but can you explain how it should be installed? I only know how to install cerebro plugins using the built in "plugins" keyword. And yours doesn't show up as an option.. only the old (and deprecated) file search is available, and that one doesn't seem to work at all on my distro (Linux Mint 19.04)

**Comment by lucky125111** [comment]
[This](https://github.com/KELiON/cerebro/issues/441#issuecomment-357970650) is the way I used. Is there any other way you can add plugins? I'm not sure how plugins keyword works, I assume it checks some repository where my plugin is not present.

**Comment by hammer** [comment]
I saw #303 was closed so I just want to be sure that searching cloud file storage systems like Dropbox and Google Drive, even when files are not present on the local file system, is included in the scope of this issue. I'd be happy to file another if not.

Fast search of both local and cloud files is the use case that brought me to Cerebro so consider me a +1 for this feature!
